### PR TITLE
Support for FAPI

### PR DESCRIPTION
**Pull request by AndreasFuchsSIT**
This is a set of tools for using the FAPI called tss2_*.
The 2nd commit just redirects to pull tpm2-tss from the PR branch on fapi-rfc.
I guess these tools can be considered mostly done. I've noticed some bad license headers
and help message (license as well) that will be fixed to align with the rest of the project.
Other than that, please feel free to review the tools and play around with.
It includes bash completion and man-pages.

I'll be force-pushing to this PR and trying to update the PR message as good as possible
until we can merge this into master and work in there. I'd give it 1-2 more weeks.

CI-Failure comes from some resource leakage in the FAPI I think. We'll be fixing this upstream.
Sorry for this.

### CODE REVIEW

File: .ci/download-deps.sh [hunk 1]
```diff
@@ -5,13 +5,36 @@ function get_deps() {
 
 	TSS_VERSION=2.3.1
 	ABRMD_VERSION=master
+	JSONC_VER=json-c-0.13.1-20180305
+
+    (apt-get update && apt-get --yes install expect) || \
+    dnf -y install expect || \
+    zypper --non-interactive install expect
```
Comment (williamcroberts) [inline .ci/download-deps.sh]: We would want to add these to the docker image, not to the script.

File: .ci/download-deps.sh [hunk 2]
```diff
@@ -5,13 +5,36 @@ function get_deps() {
...
 
 	echo "pwd starting: `pwd`"
 	pushd "$1"
+	echo "pwd clone json-c: `pwd`"
```
Comment (williamcroberts) [inline .ci/download-deps.sh]: why are we getting this one from source?

File: Makefile.am [hunk 3]
```diff
@@ -30,7 +30,54 @@ LDADD = \
...
+    tools/fapi/tss2_pcrextend \
+    tools/fapi/tss2_quote \
+    tools/fapi/tss2_pcrread \
+    tools/fapi/tss2_authorizepolicy \
+    tools/fapi/tss2_exportpolicy \
+    tools/fapi/tss2_import \
+    tools/fapi/tss2_provision \
+    tools/fapi/tss2_getrandom \
+    tools/fapi/tss2_unseal \
+    tools/fapi/tss2_writeauthorizenv
```
Comment (williamcroberts) [inline Makefile.am]: +1 for placing these into a subdir, read my mind.

File: Makefile.am [hunk 4]
```diff
@@ -436,17 +676,56 @@ CLEANFILES = $(dist_man1_MANS)
...
+	dist/bash-completion/tpm2-tools/tss2_setdescription \
+	dist/bash-completion/tpm2-tools/tss2_pcrextend \
+	dist/bash-completion/tpm2-tools/tss2_quote \
+	dist/bash-completion/tpm2-tools/tss2_pcrread \
+	dist/bash-completion/tpm2-tools/tss2_authorizepolicy \
+	dist/bash-completion/tpm2-tools/tss2_exportpolicy \
+	dist/bash-completion/tpm2-tools/tss2_provision \
+	dist/bash-completion/tpm2-tools/tss2_getrandom \
+    dist/bash-completion/tpm2-tools/tss2_unseal \
+    dist/bash-completion/tpm2-tools/tss2_writeauthorizenv
```
Comment (williamcroberts) [inline Makefile.am]: fix indenting errors.

File: tools/fapi/tss2_changeauth.c [hunk 5]
```diff
@@ -0,0 +1,69 @@
+/* SPDX-License-Identifier: BSD-2 */
+/*******************************************************************************
+ * Copyright 2019, Fraunhofer SIT sponsored by Infineon Technologies AG
+ * All rights reserved.
+ ******************************************************************************/
```
Comment (williamcroberts) [inline tools/fapi/tss2_changeauth.c]: no copy right clauses in any file, this and throughout. Those statements are superfluous.

File: .ci/download-deps.sh [hunk 6]
```diff
@@ -5,13 +5,36 @@ function get_deps() {
...
+		echo "json-c already downloaded/built/installed, skipping"
+	fi
+
 	echo "pwd clone tss: `pwd`"
 	if [ ! -d tpm2-tss ]; then
+#		git clone --depth=1 \
+#		--branch "$TSS_VERSION" https://github.com/tpm2-software/tpm2-tss.git
 		git clone --depth=1 \
-		--branch "$TSS_VERSION" https://github.com/tpm2-software/tpm2-tss.git
+		--branch "fapi-rfc" https://github.com/AndreasFuchsSIT/tpm2-tss.git
```
Comment (williamcroberts) [inline .ci/download-deps.sh]: we will wait until this is on a branch in the official tss repo

File: .ci/download-deps.sh (hunk 1 above)
Comment (williamcroberts) [reply .ci/download-deps.sh]: You can modify the .docker files here and submit a PR:
https://github.com/tpm2-software/tpm2-software-container

File: tools/fapi/tss2_createnv.c [hunk 7]
```diff
@@ -0,0 +1,93 @@
+/* SPDX-License-Identifier: BSD-2 */
```
Comment (williamcroberts) [inline tools/fapi/tss2_createnv.c]: Also the license for the project is BSD-3, can you update these.

File: tools/fapi/tss2_changeauth.c (hunk 5 above)
Comment (williamcroberts) [reply tools/fapi/tss2_changeauth.c]: For those that want these, they can place them in the LICENSE file, which you already have one.

File: Makefile.am (hunk 3 above)
Comment (AndreasFuchsSIT) [reply Makefile.am]: awesome, marked as resolved... ;-)

File: .ci/download-deps.sh (hunk 1 above)
Comment (AndreasFuchsSIT) [reply .ci/download-deps.sh]: Will do. Just have to figure out how...

File: tools/fapi/tss2_changeauth.c (hunk 5 above)
Comment (AndreasFuchsSIT) [reply tools/fapi/tss2_changeauth.c]: willdo

File: .ci/download-deps.sh (hunk 1 above)
Comment (AndreasFuchsSIT) [reply .ci/download-deps.sh]: @williamcroberts Ping (since I really have no idea about the docker stuff)

**Comment by AndreasFuchsSIT** [comment]
Ok, so this is now running against tpm2-tss master, not the private repo anymore.

Ready to merge ?

**Comment by AndreasFuchsSIT** [comment]
@diabonas Could you add libjson-c to the LGTM for this PR ?
Thanks !

**Comment by diabonas** [comment]
@AndreasFuchsSIT Done in #1877

File: dist/bash-completion/tpm2-tools/tss2_createnv [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    local cur prev words cword split
+    _init_completion -s || return
+    case $prev in
+        -!(-*)h | --help)
+            COMPREPLY=( $(compgen -W "man no-man" -- "$cur") )
+            return;;
+        -!(-*)t | --type)
+            COMPREPLY+=( '-' )
+            return;;
+        -!(-*)[psPa] | --path | --size |--policyPath | --authValue)
```
Comment (PeterHuewe) [inline dist/bash-completion/tpm2-tools/tss2_createnv]: Space between `|` and `--policyPath` required

File: dist/bash-completion/tpm2-tools/tss2_decrypt [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+_tss2_decrypt()
+{
+    local cur prev words cword split
+    _init_completion -s || return
+    case $prev in
+        -!(-*)h | --help)
+            COMPREPLY=( $(compgen -W "man no-man" -- "$cur") )
+            return;;
+        -!(-*)[kpc] | --keyPath | --plainText | --cipherText)
+            _filedir
```
Comment (PeterHuewe) [inline dist/bash-completion/tpm2-tools/tss2_decrypt]: why does `--keyPath` here has the `_filedir` completion, but many other commands (e.g. tss2_authorizepolicy) don't? -> should be consistent.

File: dist/bash-completion/tpm2-tools/tss2_createseal [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+        -!(-*)h | --help)
+            COMPREPLY=( $(compgen -W "man no-man" -- "$cur") )
+            return;;
+        -!(-*)[ptPad] | --path | --type | --policyPath | --authValue | --data)
+            return;;
+    esac
+
+    $split && return
+
+    COMPREPLY=( $(compgen -W "-h --help -v --version --password -a --path= -p --policyPath= -P --type= -t --data= -d" -- "$cur") )
```
Comment (PeterHuewe) [inline dist/bash-completion/tpm2-tools/tss2_createseal]: There is no `--password` , the manpage indicates `--authValue`

File: man/tss2_encrypt.1.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+  * **-k**, **\--keyPath**=:
+
+    Identifies the encryption key. MUST NOT be NULL.
+
+  * **-f**, **\--force**:
+
+    Force overwriting the output file.
+
+  * **-p**, **\--policyPath**:
```
Comment (PeterHuewe) [inline man/tss2_encrypt.1.md]: Most other tss2_ tools have uppercase `-P`  for policyPath

**Comment by AndreasFuchsSIT** [comment]
Agreed with all comments.

>     * If a tpm2_tool already exists the parameters and behavior should be as similar as possible in the tss2_ variant (e.g. `tpm2_getrandom` vs `tss2_getrandom`)

One remark here: there are no non-option parameters in any of the tss2_* tools. IMHO adding this for the tss2_getrandom would screw consistency here. What shall we do about that ? Proposal: Keep as is.

**Comment by PeterHuewe** [comment]
> 
> 
> Agreed with all comments.
> 
> > ```
> > * If a tpm2_tool already exists the parameters and behavior should be as similar as possible in the tss2_ variant (e.g. `tpm2_getrandom` vs `tss2_getrandom`)
> > ```
> 
> One remark here: there are no non-option parameters in any of the tss2_* tools. IMHO adding this for the tss2_getrandom would screw consistency here. What shall we do about that ? Proposal: Keep as is.

Sorry I don't understand what you mean.
Can you please elaborate?

**Comment by PeterHuewe** [comment]
tpm2_getrandom
```
    const struct option topts[] = {
        { "output",     required_argument, NULL, 'o' },
        { "force",      required_argument, NULL, 'f' },
        { "hex",        no_argument,       NULL,  0  },
    };
```
vs
tss2_getrandom
```
    struct option topts[] = {
        {"numBytes", required_argument, NULL, 'n'},
        {"force"    , no_argument      , NULL, 'f'},
        /* output file */
        {"data"   , required_argument, NULL, 'd'}
    };
```

What I mean is that I would like tss2_getrandom also use `-o` for the output.
I do not mind if there are extra options or less options, but options with similar meaning should have similar parameter.

**Comment by AndreasFuchsSIT** [comment]
> What I mean is that I would like tss2_getrandom also use `-o` for the output.
> I do not mind if there are extra options or less options, but options with similar meaning should have similar parameter.
Ok, got it. I was referring to the `-n`... `-o` makes a lot of sense...

**Comment by AndreasFuchsSIT** [comment]
Ok, all of Peter's points should be addressed now.
Looks ready for merging...

**Comment by AndreasFuchsSIT** [comment]
@idesai What do you need before you can hit merge ?

**Comment by williamcroberts** [comment]
@AndreasFuchsSIT rebase and fix the DCO issues (need signoff) and force push. This is fine.

**Comment by AndreasFuchsSIT** [comment]
Ok, the DCO had to do with the order of sign-offs (weird). The rebase was simple. Now we're waiting for the default behavior of asking the user. Once that's sQuashed, I'll trigger you for merging.

**Comment by AndreasFuchsSIT** [comment]
@williamcroberts @idesai From our side this is now ready to merge.

**Comment by AndreasFuchsSIT** [comment]
As approved by @williamcroberts

**Pull request merged**